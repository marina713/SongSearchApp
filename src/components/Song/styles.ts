import styled from 'styled-components/native';
import {ThemeColors} from '~/constants/colors';

export const Container = styled.View<{theme: ThemeColors}>`
  margin: 10px;
  border-radius: 5px;
  border-color: ${({theme}) => theme.border};
  border-width: 1px;
  padding: 5px 10px;
  align-items: center;
`;

export const TrackName = styled.Text<{theme: ThemeColors}>`
  font-size: 30px;
  color: ${({theme}) => theme.title};
  margin-bottom: 3px;
  font-weight: 600;
`;
export const AlbumCover = styled.Image`
  aspect-ratio: 1;
  width: 200px;
  border-radius: 5px;
  margin: 20px 15px 5px 0;
`;

export const Title = styled.Text`
  font-size: 23px;
`;
