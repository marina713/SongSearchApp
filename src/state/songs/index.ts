import {
  UPDATE_SONGS,
  initialState,
  SET_SORT_BY,
  SET_PLAYING_TRACK_ID,
} from './constants';
import {SongsState, Action} from './types';

export const songs = (
  state: SongsState = initialState,
  action: Action,
): SongsState => {
  switch (action.type) {
    case UPDATE_SONGS: {
      const {payload} = action;
      return {
        ...state,
        songs: payload,
        sortBy: initialState.sortBy,
      };
    }

    case SET_SORT_BY: {
      const {payload} = action;
      return {
        ...state,
        sortBy: payload,
      };
    }

    case SET_PLAYING_TRACK_ID: {
      const {payload} = action;
      return {
        ...state,
        playingTrackId: payload,
      };
    }

    default: {
      return state;
    }
  }
};
