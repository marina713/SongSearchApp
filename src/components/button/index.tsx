import * as React from 'react';

import { Container, ButtonText } from './styles';

export type Props = {
  text: string;
  onPress: () => void;
};

export const Button = React.memo(({ text, onPress }: Props) => (
  <Container onPress={onPress}>
    <ButtonText>{text}</ButtonText>
  </Container>
));
