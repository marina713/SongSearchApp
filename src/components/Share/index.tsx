import React from 'react';
import { Share, Image, TouchableOpacity } from 'react-native';
import { Song } from '~/state/songs/types';
import { Container } from './styles';

type Props = { song: Song };

export const ShareButton = ({ song }: Props) => {
  const onShare = async () => {
    try {
      const result = await Share.share({
        message: `Check out this song: ${song.trackName} by ${song.artistName}`,
      });
      if (result.action === Share.sharedAction) {
        if (result.activityType) {
          // shared with activity type of result.activityType
        } else {
          // shared
        }
      } else if (result.action === Share.dismissedAction) {
        // dismissed
      }
    } catch (error) {}
  };
  return (
    <Container>
      <TouchableOpacity onPress={onShare}>
        <Image
          source={require('~/assets/images/share.png')}
          style={{ width: 40, height: 40, margin: 5 }}
        />
      </TouchableOpacity>
    </Container>
  );
};
