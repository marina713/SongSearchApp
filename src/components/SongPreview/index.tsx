import * as React from 'react';
import { Song } from '~/state/songs/types';
import { msToHMS, formatDate } from '~/utils/time';

import {
  Container,
  TrackName,
  Title,
  AlbumCover,
  FlexCol,
  FlexRow,
  ExtraInfoContainer,
  ExtraInfoText,
  SubTitle,
} from './styles';

export type Props = {
  song: Song;
  onSongPress?: () => void;
};

export const SongPreview = React.memo(({ song, onSongPress }: Props) => {
  const price =
    song.currency === 'USD'
      ? `$${song.trackPrice}`
      : `${song.trackPrice}${song.currency}`;
  const formattedReleaseDate = song.releaseDate && formatDate(song.releaseDate);
  const releaseDate = formattedReleaseDate
    ? `Released on ${formattedReleaseDate} · `
    : '';
  const formattedLength = song.trackTimeMillis && msToHMS(song.trackTimeMillis);
  const length = formattedLength ? `Length ${formattedLength}` : '';

  return (
    <Container onPress={onSongPress}>
      <AlbumCover source={{ uri: song.artworkUrl100 }} />
      <FlexCol>
        <TrackName>{song.trackName}</TrackName>
        <Title>{`${song.artistName} · ${song.primaryGenreName}`}</Title>
        <SubTitle>{`${releaseDate}${length}`}</SubTitle>
        <FlexRow>
          <ExtraInfoContainer>
            <ExtraInfoText>{price}</ExtraInfoText>
          </ExtraInfoContainer>
        </FlexRow>
      </FlexCol>
    </Container>
  );
});
