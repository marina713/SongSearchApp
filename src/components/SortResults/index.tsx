import * as React from 'react';
import { setSortBy } from '~/state/songs/actions';
import { useSelector, useDispatch } from 'react-redux';
import { getSortBy } from '~/state/songs/selectors'

import { Container, Title, SortItem, SortItemText, SortItemsContainer, SortIcon } from './styles';
import { SortByType } from '~/state/songs/types';

type SortItemComponentProps = {
    sortBy: SortByType;
}

const SORT_BY = ['Duration', 'Genre', 'Price'];

const SortItemComponent = ({ sortBy }: SortItemComponentProps) => {
    const dispatch = useDispatch();
    const sortByState = useSelector(getSortBy);

    const isSelected = sortByState.type === sortBy;
    const source = sortByState.order === 'DESC' ?
        require('~/assets/images/sort-descending.png')
        : require('~/assets/images/sort-ascending.png');

    const onPress = () => {
        const nextValue = sortByState.type === '' || !isSelected ?
            { type: sortBy, order: 'DESC' }
            : sortByState.order === 'DESC' ?
                { type: sortBy, order: 'ASC' } : { type: '', order: '' };
        dispatch(setSortBy(nextValue))
    }

    return (
        <SortItem isSelected={isSelected} onPress={onPress}>
            <SortItemText isSelected={isSelected}>{sortBy}</SortItemText>
            {isSelected ? <SortIcon source={source} /> : null}
        </SortItem>
    )
}

export const SortResults = () => (
    <Container>
        <Title>Sort by</Title>
        <SortItemsContainer>
            {SORT_BY.map(item => <SortItemComponent key={item} sortBy={item} />)}
        </SortItemsContainer>
    </Container>
);
