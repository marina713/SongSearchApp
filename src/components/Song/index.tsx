import * as React from 'react';
import { Song as SongType } from '~/state/songs/types';
import { msToHMS, formatDate } from '~/utils/time'
import { ShareButton } from '../Share';

import {
    Container,
    TrackName,
    Title,
    AlbumCover,
    FlexRow,
    ExtraInfoContainer,
    ExtraInfoText,
    FlexRowContainer
} from './styles';

export type Props = {
    song: SongType;
};

export const Song = React.memo(({ song }: Props) => {
    const formattedReleaseDate = song.releaseDate && `Released ${formatDate(song.releaseDate)}`;
    const formattedLength = song.trackTimeMillis && `Length ${msToHMS(song.trackTimeMillis)}`;
    const price = song.currency === 'USD' ? `$${song.trackPrice}` : `${song.trackPrice}${song.currency}`;
    const genre = song.primaryGenreName || '';

    const extraInfo = [genre, formattedReleaseDate, formattedLength, price];

    return (
        <Container>
            <TrackName>{song.trackName}</TrackName>
            <Title>{song.artistName}</Title>
            <AlbumCover source={{ uri: song.artworkUrl100 }} />
            <FlexRowContainer>
                <ShareButton song={song} />
                <FlexRow>
                    {extraInfo.map(item => (item ?
                        <ExtraInfoContainer key={item}>
                            <ExtraInfoText>{item}</ExtraInfoText>
                        </ExtraInfoContainer> : null))}
                </FlexRow>
            </FlexRowContainer>
        </Container>
    )
});
