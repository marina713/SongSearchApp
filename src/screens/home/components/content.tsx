import * as React from 'react';

import { FlatList, Subtitle, NotFoundContainer, Lottie } from '../styles';
import { SongPreview } from '~/components/SongPreview';
import { SortResults } from '~/components/SortResults';
import { Song } from '~/state/songs/types';

type HomeContentProps = {
  songs: Song[];
  onSongPress: (trackId: number) => void;
  isTyping: boolean;
  text: string;
};

export const HomeContentComponent = React.memo(
  ({ songs, onSongPress, isTyping, text }: HomeContentProps) =>
    songs.length > 0 ? (
      <>
        <SortResults />
        <FlatList
          data={songs}
          renderItem={({ item }: { item: any }) => (
            <SongPreview
              song={item}
              onSongPress={() => onSongPress(item.trackId)}
            />
          )}
          keyExtractor={(item: any, index) => `${item.trackId}-${index}`}
        />
      </>
    ) : isTyping || !text ? null : (
      <NotFoundContainer>
        <Lottie source={require('~/assets/lotties/not-found.json')} autoPlay />
        <Subtitle> No music found, try again!</Subtitle>
      </NotFoundContainer>
    ),
);
