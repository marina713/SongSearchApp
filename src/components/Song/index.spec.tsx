import * as React from 'react';
import renderer from 'react-test-renderer';
import { MOCK_SONGS } from '~/state/songs/mocks';

import { Song, Props } from '~/components/Song';

jest.mock('./styles', () => ({
  Container: 'Container',
  ButtonText: 'ButtonText',
  TrackName: 'TrackName',
  Title: 'Title',
  AlbumCover: 'AlbumCover',
  ExtraInfoContainer: 'ExtraInfoContainer',
  ExtraInfoText: 'ExtraInfoText',
  FlexRowContainer: 'FlexRowContainer',
}));

const initialProps = {
  song: MOCK_SONGS[0],
};

const testComponent = (props?: Props) => {
  return renderer.create(<Song {...initialProps} {...props} />);
};

describe('<Song />', () => {
  it('renders correctly with default props', () => {
    const tree = testComponent().toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with custom Song', () => {
    const tree = testComponent({
      ...initialProps,
      song: MOCK_SONGS[2],
    }).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
