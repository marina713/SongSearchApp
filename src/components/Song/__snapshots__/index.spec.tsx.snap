// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Song /> renders correctly with custom Song 1`] = `
<Container>
  <TrackName>
    Come As You Are
  </TrackName>
  <Title>
    Nirvana
  </Title>
  <AlbumCover
    source={
      Object {
        "uri": "https://is3-ssl.mzstatic.com/image/thumb/Music125/v4/8d/58/a9/8d58a928-0663-71f7-4e5c-d2bd54a124fb/source/100x100bb.jpg",
      }
    }
  />
  <FlexRowContainer>
    <ExtraInfoContainer>
      <ExtraInfoText>
        Alternative
      </ExtraInfoText>
    </ExtraInfoContainer>
    <ExtraInfoContainer>
      <ExtraInfoText>
        Length 3:39
      </ExtraInfoText>
    </ExtraInfoContainer>
    <ExtraInfoContainer>
      <ExtraInfoText>
        Released 24/9/1991
      </ExtraInfoText>
    </ExtraInfoContainer>
    <ExtraInfoContainer>
      <ExtraInfoText>
        $1.29
      </ExtraInfoText>
    </ExtraInfoContainer>
  </FlexRowContainer>
</Container>
`;

exports[`<Song /> renders correctly with default props 1`] = `
<Container>
  <TrackName>
    Nirvana
  </TrackName>
  <Title>
    Sam Smith
  </Title>
  <AlbumCover
    source={
      Object {
        "uri": "https://is4-ssl.mzstatic.com/image/thumb/Music114/v4/e4/f9/59/e4f959ff-e758-ff29-4b26-39ae222489c8/source/100x100bb.jpg",
      }
    }
  />
  <FlexRowContainer>
    <ExtraInfoContainer>
      <ExtraInfoText>
        Pop
      </ExtraInfoText>
    </ExtraInfoContainer>
    <ExtraInfoContainer>
      <ExtraInfoText>
        Length 3:17
      </ExtraInfoText>
    </ExtraInfoContainer>
    <ExtraInfoContainer>
      <ExtraInfoText>
        Released 1/1/2013
      </ExtraInfoText>
    </ExtraInfoContainer>
    <ExtraInfoContainer>
      <ExtraInfoText>
        $0.69
      </ExtraInfoText>
    </ExtraInfoContainer>
  </FlexRowContainer>
</Container>
`;
